rnd/autogpt_server/autogpt_server/data/graph.py

```diff
@@ -126,65 +165,100 @@ def sanitize(name):
 
-        # Check if all links are connected compatible pin data type.
+        def is_input_output_block(nid: str) -> bool:
+            bid = node_map[nid].block_id
+            b = get_block(bid)
+            return isinstance(b, InputBlock) or isinstance(b, OutputBlock)
+
+        # subgraphs: all nodes in subgraph must be present in the graph.
+        for subgraph_id, node_ids in self.subgraphs.items():
+            for node_id in node_ids:
+                if node_id not in node_map:
+                    raise ValueError(f"Subgraph {subgraph_id}'s node {node_id} invalid")
+        subgraph_map = self.subgraph_map
+
+        # Links: links are connected and the connected pin data type are compatible.
         for link in self.links:
-            source_id = link.source_id
-            sink_id = link.sink_id
-            suffix = f"Link {source_id}<->{sink_id}"
-
-            source_node = next((v for v in self.nodes if v.id == source_id), None)
-            if not source_node:
-                raise ValueError(f"{suffix}, {source_id} is invalid node.")
-            sink_node = next((v for v in self.nodes if v.id == sink_id), None)
-            if not sink_node:
-                raise ValueError(f"{suffix}, {sink_id} is invalid node.")
-
-            source_block = get_block(source_node.block_id)
-            if not source_block:
-                raise ValueError(f"{suffix}, {source_node.block_id} is invalid block.")
-            sink_block = get_block(sink_node.block_id)
-            if not sink_block:
-                raise ValueError(f"{suffix}, {sink_node.block_id} is invalid block.")
-
-            source_name = sanitize(link.source_name)
-            if source_name not in source_block.output_schema.get_fields():
-                raise ValueError(f"{suffix}, `{source_name}` is invalid output pin.")
-            sink_name = sanitize(link.sink_name)
-            if sink_name not in sink_block.input_schema.get_fields():
-                raise ValueError(f"{suffix}, `{sink_name}` is invalid input pin.")
+            source = (link.source_id, link.source_name)
+            sink = (link.sink_id, link.sink_name)
+            suffix = f"Link {source} <-> {sink}"
+
+            for i, (node_id, name) in enumerate([source, sink]):
+                node = node_map.get(node_id)
+                if not node:
+                    raise ValueError(f"{suffix}, {node_id} is invalid node.")
+
+                block = get_block(node.block_id)
+                if not block:
+                    raise ValueError(f"{suffix}, {node.block_id} is invalid block.")
+
+                sanitized_name = sanitize(name)
+                if i == 0:
+                    fields = block.output_schema.get_fields()
+                else:
+                    fields = block.input_schema.get_fields()
+                if sanitized_name not in fields:
+                    raise ValueError(f"{suffix}, `{name}` invalid, fields: {fields}")
+
+            if (
```

Comment:
Basically this is the main change: 
You can now group nodes into a subgraph, and you are not allowed to connect nodes from the different subgraph.